tensorflow_lite_support/ios/task/vision/sources/TFLObjectDetector.m

```diff
@@ -0,0 +1,123 @@
+
+@interface TFLObjectDetector ()
+/** ObjectDetector backed by C API */
+@property(nonatomic) TfLiteObjectDetector *objectDetector;
+@end
+
+@implementation TFLObjectDetectorOptions
+@synthesize baseOptions;
+@synthesize classificationOptions;
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    self.baseOptions = [[TFLBaseOptions alloc] init];
+    self.classificationOptions = [[TFLClassificationOptions alloc] init];
+  }
+  return self;
+}
+
+- (nullable instancetype)initWithModelPath:(nonnull NSString *)modelPath {
+  self = [self init];
+  if (self) {
+    self.baseOptions.modelFile.filePath = modelPath;
+  }
+  return self;
+}
+
+@end
+
+@implementation TFLObjectDetector
+- (void)dealloc {
+  TfLiteObjectDetectorDelete(_objectDetector);
+}
+
+- (instancetype)initWithObjectDetector:(TfLiteObjectDetector *)objectDetector {
+  self = [super init];
+  if (self) {
+    _objectDetector = objectDetector;
+  }
+  return self;
+}
+
++ (nullable instancetype)objectDetectorWithOptions:(nonnull TFLObjectDetectorOptions *)options
+                                             error:(NSError **)error {
+  TfLiteObjectDetectorOptions cOptions = TfLiteObjectDetectorOptionsCreate();
+  if (![options.classificationOptions
+          copyClassificationOptionsToCClassificationOptions:&(cOptions.classification_options)
+                                                      error:error])
+    return nil;
+
+  [options.baseOptions copyBaseOptionsToCBaseOptions:&(cOptions.base_options)];
+
+  TfLiteSupportError *createObjectDetectorError = nil;
+  TfLiteObjectDetector *objectDetector =
+      TfLiteObjectDetectorFromOptions(&cOptions, &createObjectDetectorError);
+
+  [options.classificationOptions
+      deleteCStringArraysOfClassificationOptions:&(cOptions.classification_options)];
+
+  if (!objectDetector) {
+    [TFLCommonUtils errorFromTfLiteSupportError:createObjectDetectorError error:error];
+    TfLiteSupportErrorDelete(createObjectDetectorError);
+    return nil;
+  }
+
+  return [[TFLObjectDetector alloc] initWithObjectDetector:objectDetector];
+}
+
+- (nullable TFLDetectionResult *)detectWithGMLImage:(GMLImage *)image
+                                              error:(NSError *_Nullable *)error {
+  TfLiteFrameBuffer *cFrameBuffer = [GMLImageUtils cFrameBufferFromGMLImage:image error:error];
+
+  if (!cFrameBuffer) {
+    return nil;
+  }
+
+  TfLiteSupportError *detectError = nil;
+  TfLiteDetectionResult *cDetectionResult =
+      TfLiteObjectDetectorDetect(_objectDetector, cFrameBuffer, &detectError);
+
+  free(cFrameBuffer->buffer);
+  cFrameBuffer->buffer = nil;
+
+  free(cFrameBuffer);
+  cFrameBuffer = nil;
+
+  if (!cDetectionResult) {
+    [TFLCommonUtils errorFromTfLiteSupportError:detectError error:error];
+    TfLiteSupportErrorDelete(detectError);
+    return nil;
+  }
+
+  TFLDetectionResult *detectionResult =
+      [TFLDetectionResult detectionResultFromCDetectionResults:cDetectionResult];
```

Comment:
detectionResultFromCDetectionResult was intended to be similar to naming conventions like [UIImage imageWithCGImage:cg_image]. I think 'with' is more common than from in native iOS methods. 'copyFromCResults' can also be used but looses the commonly used ObjC naming standards. @khanhlvg Could you give us your insight?

Reply:
Let's use `detectionResultWithCDetectionResults:cDetectionResult` to align with Apple's ObjC convention.

Reply:
My concern here (and same for a few other similar comments) is that the method names are too verbose and lengthy, which add frictions to code readability. I found this [Apple Naming rule](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html#//apple_ref/doc/uid/20001282-BCIGIJJF) that might be helpful here. There's a few factory method example, i.e. `initWithFrame`. So we could have `initWithCDetectionResults`.
Also "For methods that represent actions an object takes, start the name with a verb". A few examples are `invokeWithTarget`, `selectTabViewItem`.

Reply:
init is prefixed only to initialisation methods usually called like `[[UIImage alloc] initWithName:name]`. This one is a class method (static). copy is not a keyword used when the class method returns object of same class for eg: `[UIImage imageWithCGImage:cg_image]`. They omit the prefix for the class and use the the rest to begin the method name. And for the class from which it is initialised which usually appears before the parameter, the same is done so that the name flows like a sentence and is very explanatory.  By convention, long method names are not an issue in Obj C. But we can name it any way we want. So if you want it shorter we can have something like.  `[TFLDetectionResult detectionResultFromCResult:cDetectionResult]`;  @khanhlvg Does this sound good ?

Reply:
@priankakariatyml Can you use initialization methods rather than class methods?

So for example, `detectionResult = [[TFLDetectionResult alloc] initWithCDetectionResult: cDetectionResult]` instead of `detectionResult = [TFLDetectionResult detectionResultWithCDetectionResult: cDetectionResult]`.

Reply:
@lu-wang-g FYI, GTMHttpFetcher also use both init methods and class methods to initialize objects. 
https://github.com/google/gtm-http-fetcher/blob/a3e318b101410422bd510f9afd6bc9c28ec87019/Source/GTMHTTPFetcher.h#L577

In this case, I don't have a strong preference over with one to use. The init method results in a shorter method name, but as you need to "alloc" the object, the total number of character requires for an initialization call doesn't change much.

Reply:
Thanks for the discussions here, Prianka and Khanh! `detectionResultFromCResult` sounds good to me. In fact, I dug more in the naming convention for obj-c, especially [factory method](https://developer.apple.com/library/archive/documentation/General/Conceptual/CocoaEncyclopedia/ClassFactoryMethods/ClassFactoryMethods.html). It recommends the pattern of `<ClassName>With<InputType>`, which is aligned with Prianka's proposal.

Reply:
I have renamed all methods accordingly